### PR TITLE
feat(api): Add method to consume metadata endpoint

### PR DESCRIPTION
**Pull request by spencerwilson**
- `SourcePath`: add method `qualified_name`
- Add method to consume the new `GET /sources/{}/metadata` endpoint

## Test plan

See upstack PR.

### CODE REVIEW

**Comment by spencerwilson** [comment]
Current dependencies on/for this PR:
* main
  * **PR #192** <a href="https://app.graphite.dev/github/pr/patch-tech/dpm/192?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #193** <a href="https://app.graphite.dev/github/pr/patch-tech/dpm/193?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
      * **PR #194** <a href="https://app.graphite.dev/github/pr/patch-tech/dpm/194?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/patch-tech/dpm/193?utm_source=stack-comment).

File: src/api.rs [hunk 1]
```diff
@@ -106,6 +107,22 @@ impl Client {
         Ok(serde_json::from_str(&body)?)
     }
 
+    #[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (spencerwilson) [reply src/api.rs:110]: rustc (or clippy; I can't recall) warns that this method is not referenced anywhere in the tree, that's what this attribute is squelching. The method is referenced in the next-upstack PR, so the attribute is removed in that PR.

File: src/api.rs [hunk 2]
```diff
@@ -234,6 +251,70 @@ impl Source {
...
+    /// [Decimal fraction](https://en.wikipedia.org/wiki/Decimal#Decimal_fractions).
+    Number,
+    /// Date in the proleptic Gregorian calendar.
+    Date,
+    /// Triple of hours, minutes, and seconds.
+    Time,
+    /// A combination of `Date` and `Time`, referring to an instant on the UTC
+    /// time scale.
+    DateTime,
+    /// A list of values. May or may not be homogeneous.
```
Comment (spencerwilson) [inline src/api.rs:279]: > May or may not be homogeneous

Snowflake's `ARRAY` data type is heterogeneous, BigQuery's and Databricks's are homogeneous. Arrow's equivalent, [List](https://docs.rs/arrow/49.0.0/arrow/datatypes/enum.DataType.html#variant.List), is homogeneous as well.

File: src/descriptor/data_package.rs [hunk 3]
```diff
@@ -92,8 +114,172 @@ pub struct DataResource {
...
+    type Error = String;
+
+    fn try_from(value: api::TableSchema) -> std::result::Result<Self, Self::Error> {
+        let fields: Vec<TableSchemaField> = value
+            .fields
+            .into_iter()
+            .filter_map(|f| {
+                let field_name = f.name.clone();
+
+                match f.try_into() {
```
Comment (ajmasci) [inline src/descriptor/data_package.rs:151]: Trying to understand under what circumstances would one go down the `Err` arm?

File: src/descriptor/data_package.rs [hunk 4]
```diff
@@ -92,8 +114,172 @@ pub struct DataResource {
...
+        Ok(match dpm_beta_type {
+            api::DpmBetaType::String => TableSchemaField::StringField {
+                constraints: Some(base_constraints),
+                description: None,
+                example: None,
+                format: StringFieldFormat::Default,
+                name: value.name,
+                rdf_type: None,
+                title: None,
+                type_: Some(StringFieldType::String),
```
Comment (ajmasci) [inline src/descriptor/data_package.rs:219]: Why does this `type_` take an optional value, while the others below don't?

**Review by ajmasci — APPROVED** [review]
A few questions for your consideration, but all non-blocking.

File: src/descriptor/data_package.rs (hunk 3 above)
Comment (spencerwilson) [reply src/descriptor/data_package.rs:151]: See L199 and L202. Those are “Patch doesn’t support this source type at all” and “this CLI version doesn’t support this source type”.

File: src/descriptor/data_package.rs (hunk 4 above)
Comment (spencerwilson) [reply src/descriptor/data_package.rs:219]: It comes from the `type` not being a required field on string fields in the JSON schema: https://github.com/patch-tech/dpm/blob/38e9aa4c89149f9c1c3d76353c93cfa059fcb11d/src/descriptor/table_schema.json#L24-L26

I fixed the same quirk when it affected number fields upstream: https://github.com/frictionlessdata/specs/pull/829. Guess I overlooked the same issue for string fields!

**Pull request merged**